nats/src/main/java/io/micronaut/nats/intercept/NatsConsumerAdvice.java

```diff
@@ -60,101 +79,133 @@ public class NatsConsumerAdvice implements ExecutableMethodProcessor<Subject>, A
 
-    private final Map<Dispatcher, String> consumerDispatchers = new ConcurrentHashMap<>();
+    private final ApplicationConfiguration applicationConfiguration;
+
+    private final Map<String, ConsumerState> consumers = new ConcurrentHashMap<>();
+
+    private final AtomicInteger clientIdGenerator = new AtomicInteger(10);
 
     /**
      * Default constructor.
-     * @param beanContext      The bean context
-     * @param binderRegistry   The registry to bind arguments to the method
-     * @param serDesRegistry   The serialization/deserialization registry
-     * @param exceptionHandler The exception handler to use if the consumer isn't a handler
+     *
+     * @param beanContext              The bean context
+     * @param binderRegistry           The registry to bind arguments to the method
+     * @param serDesRegistry           The serialization/deserialization registry
+     * @param exceptionHandler         The exception handler to use if the consumer isn't a handler
+     * @param applicationConfiguration The application configuration
      */
     public NatsConsumerAdvice(BeanContext beanContext, NatsBinderRegistry binderRegistry,
-            NatsMessageSerDesRegistry serDesRegistry, NatsListenerExceptionHandler exceptionHandler) {
+        NatsMessageSerDesRegistry serDesRegistry, NatsListenerExceptionHandler exceptionHandler,
+        ApplicationConfiguration applicationConfiguration) {
         this.beanContext = beanContext;
         this.binderRegistry = binderRegistry;
         this.serDesRegistry = serDesRegistry;
         this.exceptionHandler = exceptionHandler;
+        this.applicationConfiguration = applicationConfiguration;
     }
 
     @Override
     public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> method) {
-        if (method.hasAnnotation(NatsListener.class)) {
-            AnnotationValue<Subject> subjectAnn = method.getAnnotation(Subject.class);
+        AnnotationValue<NatsListener> listenerAnnotation = method.getAnnotation(NatsListener.class);
+        List<AnnotationValue<Subject>> subjectAnnotations =
+            method.getDeclaredAnnotationValuesByType(Subject.class);
 
-            if (subjectAnn != null) {
-                String subject = subjectAnn.getRequiredValue(String.class);
+        if (listenerAnnotation == null || subjectAnnotations.isEmpty()) {
+            // Nothing to do if no consumer or subject annotation is present
+            return;
+        }
 
-                String connectionName =
-                        method.findAnnotation(NatsConnection.class).flatMap(conn -> conn.get("connection", String.class))
-                                .orElse(NatsConnection.DEFAULT_CONNECTION);
+        String connectionName =
+            method.findAnnotation(NatsConnection.class)
+                  .flatMap(conn -> conn.get("connection", String.class))
+                  .orElse(NatsConnection.DEFAULT_CONNECTION);
```

Comment:
you can use `method.stringValue(NatsConnection.class, "connection").orElse(NatsConnection.DEFAULT_CONNECTION)` which is more efficient

Reply:
done 😄 I like more efficient